### PR TITLE
New client service - PUT/POST - update or create client

### PR DESCRIPTION
**Pull request by mabartos**
- Closes #46785
- Implements the PUT/POST logic with `createOrUpdate`
- Use roles handling from the DefaultClientService in unified form
- Create RolesService for handling roles
- Separate logic in the `/services` to be able to use it on other places
- Fix bug of mapping protocol to the existing model
- Temporary client creation for old -> new client representations for client policies
- Also ran with the https://github.com/keycloak/keycloak/pull/47282 and tests passed

### CODE REVIEW

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 1]
```diff
@@ -115,4 +151,93 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+                    throw new ServiceException(r.getAllErrorsAsString(), Response.Status.BAD_REQUEST);
+                });
+                session.clientPolicy().triggerOnEvent(new AdminClientRegisteredContext(model, permissions.adminAuth()));
+            }
+        } catch (ClientPolicyException e) {
+            throw new ServiceException(e.getErrorDetail(), Response.Status.BAD_REQUEST);
+        }
+
+        if (model == null) {
+            throw new ServiceException("Cannot create/update client");
```
[the remark this answers is not in the document]
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: Irrelevant... we want to throw 400(default)

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 2]
```diff
@@ -115,4 +151,109 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+                        });
+
+                        session.clientPolicy().triggerOnEvent(new AdminClientUpdatedContext(proposedRepresentation, model, permissions.adminAuth()));
+                    }
+                }
+            } else {
+                // Check permissions, execute validations and trigger client policies
+                permissions.clients().requireManage();
+                validator.validate(client, strategy.getValidationGroup(), Default.class);
+                session.clientPolicy().triggerOnEvent(new AdminClientRegisterContext(oldRepresentation(clientId, client.getProtocol()), permissions.adminAuth()));
```
Comment (mabartos) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: TODO - need to check what fields are generally expected for the registerContext.

Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: Done - we also need to provide the full old representation.

**Comment by vmuzikar** [comment]
@shawkins @Pepo48 Could you please review as well?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 3]
```diff
@@ -115,4 +151,99 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+                        session.clientPolicy().triggerOnEvent(new AdminClientUpdateContext(proposedRepresentation, model, permissions.adminAuth()));
+
+                        // Update model
+                        model = mapper.toModel(client, model);
+
+                        // Validate the fully populated model
+                        ValidationUtil.validateClient(session, model, false, r -> {
+                            session.getTransactionManager().setRollbackOnly();
+                            throw new ServiceException(r.getAllErrorsAsString(), Response.Status.BAD_REQUEST);
+                        });
```
Comment (vmuzikar) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:179]: This is not ideal, we're combining two validation mechanisms – the new one based on Jakarta Validation API and our old custom one. Also, it might be confusing to the user that the error message might contain model-specific fields that are not present in the representation. I'm afraid we'll need to rewrite all the validations to the new Jakarta Validations. But that's a follow-up, and possibly even out of scope of the MVP. @mabartos Do we have a tracker for this already?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 4]
```diff
@@ -115,4 +151,99 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+        fireAdminEvent(alreadyExists ? OperationType.UPDATE : OperationType.CREATE, mapper.fromModel(model));
+        return new CreateOrUpdateResult(mapper.fromModel(model), !alreadyExists);
+    }
+
+    /**
+     * Creates a temporary client to convert BaseClientRepresentation to ClientRepresentation.
+     * Required because client policy contexts expect ClientRepresentation (v1), but there's no
+     * direct converter from BaseClientRepresentation (v2 API). The temp client is immediately removed.
+     */
+    private ClientRepresentation getProposedOldRepresentation(RealmModel realm, BaseClientRepresentation client, ClientModelMapper mapper) {
```
Comment (vmuzikar) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:236]: As discussed offline, this is not ideal. @mabartos Can you please create a follow-up issue to look into how we could resolve that so we don't need to create fake clients?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.keycloak.services;
+
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RoleContainerModel;
+import org.keycloak.services.resources.admin.AdminEventBuilder;
+import org.keycloak.services.resources.admin.RoleContainerResource;
+import org.keycloak.services.resources.admin.fgap.AdminPermissionEvaluator;
+
+public class RolesService {
```
Comment (vmuzikar) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java:10]: This is still using v1. Do we plan to replace it in a follow-up?

**Review by vmuzikar — DISMISSED** [review]
@mabartos Thank you for the changes. In general LGTM, but not merging yet as I'd like to get some additional reviews.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 3 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:179]: Yes, it's not ideal, but the ClientValidationProvider validates the ClientModel instead of the representation now. It contains some additional checks that we haven't covered yet in our Jakarta validators. 

It might be a wider topic to provide a meaningful alternative to this validationProvider and using the JakartaValidator. It should be for the 26.7.0 release. 

I think this can be closed:
- https://github.com/keycloak/keycloak/issues/43296

as the basic validations are introduced on the representation level. 

The other validations that are part of the `ClientValidationProvider` need to be carefully translated to the Jakarta validators. 

@vmuzikar Do we have a tracker for improvements that we plan for the 26.7.0 release?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java (hunk 5 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java:10]: @vmuzikar Not really, at this moment. I think we can still use the v1 for resources like realm, users, roles, etc. We just want to get rid of client resources as we want to be able to use the Client representations v2, provide declarative approach,... And as the structure of the roles has not changed, there might not be a rush to create another API for roles - we just manually handle the declarative behavior on the clients side. And as the Roles models/entities are a separate thing from the client, there's no better solution for that IMO.

**Comment by mabartos** [comment]
We should wait for the: 
- https://github.com/keycloak/keycloak/pull/47282

Or we can make some additional changes in follow-ups - I don't have strong preferences.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 4 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:236]: Created spike: 
- https://github.com/keycloak/keycloak/issues/47576

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 3 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:179]: Created spike issue: 
- https://github.com/keycloak/keycloak/issues/47577

**Comment by mabartos** [comment]
@vmuzikar Your comments should be addressed now.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java (hunk 5 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/RolesService.java:10]: As discussed, there's the follow-up for roles:
- https://github.com/keycloak/keycloak/issues/47631

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 4 above)
Comment (Pepo48) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:236]: Could we directly maps v2 fields to a v1 ClientRepresentation in memory? (see suggestion below)

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 6]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+        ClientModel tempModel = mapper.toModel(client, realm.addClient(tempId));
+        try {
+            var proposedRepresentation = ModelToRepresentation.toRepresentation(tempModel, session);
+            proposedRepresentation.setClientId(client.getClientId());
+            return proposedRepresentation;
+        } finally {
+            realm.removeClient(tempModel.getId());
+        }
+    }
 }
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:247]:
```suggestion
     /**
     * Converts a v2 BaseClientRepresentation to a v1 ClientRepresentation without creating
     * any database entities. This is needed because client policy contexts expect ClientRepresentation (v1).
     * <p>
     * For more details, see the <a href="https://github.com/keycloak/keycloak/issues/47576">keycloak#47576</a>.
     */
    private ClientRepresentation toProposedV1Representation(BaseClientRepresentation client) {
        ClientRepresentation rep = new ClientRepresentation();
        rep.setClientId(client.getClientId());
        rep.setName(client.getDisplayName());
        rep.setDescription(client.getDescription());
        rep.setEnabled(Boolean.TRUE.equals(client.getEnabled()));
        rep.setBaseUrl(client.getAppUrl());
        rep.setRootUrl(client.getAppUrl());
        rep.setProtocol(client.getProtocol());
        rep.setRedirectUris(client.getRedirectUris() != null ? new ArrayList<>(client.getRedirectUris()) : null);

        if (client instanceof OIDCClientRepresentation oidc) {
            boolean isPublic = oidc.getAuth() == null;
            rep.setPublicClient(isPublic);
            if (!isPublic) {
                rep.setClientAuthenticatorType(oidc.getAuth().getMethod());
                rep.setSecret(oidc.getAuth().getSecret());
            }

            Set<OIDCClientRepresentation.Flow> flows = Optional.ofNullable(oidc.getLoginFlows()).orElse(Collections.emptySet());
            rep.setStandardFlowEnabled(flows.contains(OIDCClientRepresentation.Flow.STANDARD));
            rep.setImplicitFlowEnabled(flows.contains(OIDCClientRepresentation.Flow.IMPLICIT));
            rep.setDirectAccessGrantsEnabled(flows.contains(OIDCClientRepresentation.Flow.DIRECT_GRANT));
            rep.setServiceAccountsEnabled(flows.contains(OIDCClientRepresentation.Flow.SERVICE_ACCOUNT));

            if (oidc.getWebOrigins() != null) {
                rep.setWebOrigins(new ArrayList<>(oidc.getWebOrigins()));
            }
        }

        return rep;
    }
}
```

This should preserve the exact semantics for client policy executors.
I also did some checks with this fix in place, all fields seems to be mapped and `ClientApiV2Test`, `ClientPoliciesV2Test`, `PatchClientValidationTest`, `PostClientValidationTest` and `PutClientValidationTest` passed.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 7]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+
+        try {
+            if (alreadyExists) {
+                switch (strategy) {
+                    case ONLY_CREATE -> throw new ServiceException("Client already exists", Response.Status.CONFLICT);
+                    case PUT, PATCH -> {
+                        // Check permissions, execute validations and trigger client policies
+                        permissions.clients().requireConfigure(model);
+                        validator.validate(client, strategy.getValidationGroup(), Default.class);
+                        var proposedRepresentation = getProposedOldRepresentation(realm, client, mapper);
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:169]:
```suggestion
                        var proposedRepresentation = toProposedV1Representation(client);
```

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 8]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+                        });
+
+                        session.clientPolicy().triggerOnEvent(new AdminClientUpdatedContext(proposedRepresentation, model, permissions.adminAuth()));
+                    }
+                }
+            } else {
+                // Check permissions, execute validations and trigger client policies
+                permissions.clients().requireManage();
+                validator.validate(client, strategy.getValidationGroup(), Default.class);
+                var proposedRepresentation = getProposedOldRepresentation(realm, client, mapper);
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:188]:
```suggestion
                var proposedRepresentation = toProposedV1Representation(client);
```

File: rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java:70]:
```suggestion
        existingModel.setProtocol(rep.getProtocol());
        existingModel.setEnabled(Boolean.TRUE.equals(rep.getEnabled()));
        existingModel.setClientId(rep.getClientId());
        existingModel.setDescription(rep.getDescription());
        existingModel.setName(rep.getDisplayName());
        existingModel.setBaseUrl(rep.getAppUrl());
        existingModel.setRedirectUris(new HashSet<>(rep.getRedirectUris()));
        // Roles are not handled here

        toModelSpecific((T) rep, existingModel);

        return existingModel;
    }
```

I think this is just a leftover the single-arg `toModel(rep)` was used in the original mapper commit (`421835c1e6`) but was abandoned by the time `DefaultClientService` introduction (see the comment on the  line 165 that says that we can't use `mapper.toModel(client)` directly for creation).

Thus, I believe createClientModel() here and the `default toModel(rep)` in `RepModelMapper` can be cleaned up.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 9]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+                });
+                session.clientPolicy().triggerOnEvent(new AdminClientRegisteredContext(model, permissions.adminAuth()));
+            }
+        } catch (ClientPolicyException e) {
+            throw new ServiceException(e.getErrorDetail(), Response.Status.BAD_REQUEST);
+        }
+
+        if (model == null) {
+            throw new ServiceException("Cannot create/update client");
+        }
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: model is always non-null at this point, no?

**Review by Pepo48 — CHANGES_REQUESTED** [review]
@mabartos thanks for the PR! I added a couple of suggestions below.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 10]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
         }
     }
+
+    private CreateOrUpdateResult createOrUpdate(RealmModel realm, String clientId, BaseClientRepresentation client, CreateOrUpdateStrategy strategy) throws ServiceException {
+        validateUnknownFields(client);
+        if (!strategy.equals(CreateOrUpdateStrategy.ONLY_CREATE)) {
+            assertSameClientIds(clientId, client.getClientId());
+        }
+
+        ClientModel model = avoidClientIdPhishing(realm.getClientByClientId(clientId)).orElse(null);
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: I know it's probably an edge case (since even in the tests we always grant both `VIEW` and `MANAGE` together), but  about scenario, when a user has `MANAGE` permission but no realm roles (`canManage() = true`, `canList() = false`). I suspect that for MANAGE-only grant, the phishing check would block a legitimate creation with 403.

Regardless of it, isn't the phishing check is semantically wrong for create/update? Like there is no risk of an information leak during these operations, these are expected conditions and `requireManage()` and/or `requireConfigure()` should cover it in terms of authorization.
```suggestion
        ClientModel model = realm.getClientByClientId(clientId);
```

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java [hunk 11]
```diff
@@ -115,4 +147,105 @@ public void deleteClient(RealmModel realm, String clientId) throws ServiceExcept
...
+        var clientRoles = rolesService.resource(model);
+        handleRoles(clientRoles, client.getRoles());
+
+        // OIDC specific
+        if (client instanceof OIDCClientRepresentation oidcClient) {
+            handleServiceAccount(clientRoles, rolesService.resource(realm), model, oidcClient);
+        }
+
+        fireAdminEvent(alreadyExists ? OperationType.UPDATE : OperationType.CREATE, mapper.fromModel(model));
+        return new CreateOrUpdateResult(mapper.fromModel(model), !alreadyExists);
```
Comment (Pepo48) [inline rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:226]: `mapper.fromModel(model)` called twice:
```suggestion
        var result = mapper.fromModel(model);
        fireAdminEvent(alreadyExists ? OperationType.UPDATE : OperationType.CREATE, result);
        return new CreateOrUpdateResult(result, !alreadyExists);
```

**Review by Pepo48 — CHANGES_REQUESTED** [review]

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 10 above)
Comment (vmuzikar) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: @Pepo48 Were you able to reproduce? At least in FGAP v1 it seems view permissions also consider manage:
https://github.com/keycloak/keycloak/blob/a2c1055f8df21b53b969f922fb9a044a76d22604/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissions.java#L295

**Comment by Pepo48** [comment]
During the review I noticed one more thing. In `DefaultClientService` class, `handleServiceAccount` method gathers both client and realm roles and then passes all of them to `addRealmRoleMappings(),` which internally calls `realm.getRole(name)`. That returns `null` (`NotFoundException`) for client roles.

The behaviour was introduced in #44587, this PR just inherits it.

Same issue exists for `deleteRealmRoleMappings()`. The mapper (`OIDCClientModelMapper.getServiceAccountRoles()`) also returns all roles including client roles, so client role names would appear in GET but fail on PUT.

I think that for now it should be scoped to realm roles only until the roles v2 API #47631 lands.

To reproduce it:
```
/.kc.sh start-dev --features=client-admin-api
./kcadm.sh --v2 config credentials --server http://localhost:8080 --realm master --user admin --password admin
./kcadm.sh --v2 client create oidc --client-id client-bug --roles my-client-role --login-flows SERVICE_ACCOUNT --service-account-roles default-roles-master --auth-method client_secret
./kcadm.sh --v2 client update oidc client-bug --client-id client-bug --roles my-client-role --login-flows SERVICE_ACCOUNT --service-account-roles my-client-role --auth-method client_secret
```

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 10 above)
Comment (Pepo48) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: Yes, I tested it during the review. As I understand it, in V1, [`canView()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissions.java#L294-L296) checks for the `MANAGE_CLIENTS` role - not the FGAP manage scope. And since [`canManage()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissions.java#L283-L285) in V1 is purely role-based (`MANAGE_CLIENTS` role), any user who passes `requireManage()` automatically passes [`canList()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissions.java#L248) too. The phishing check never blocks.

V2 introduced a new path: a user can pass [`requireManage()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissionsV2.java#L72-L78) via FGAP manage scope without the `MANAGE_CLIENTS` role. But [`canView()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissionsV2.java#L90-L96) in V2 only checks for the FGAP view scope - manage doesn't imply view. So [`canList()`](https://github.com/keycloak/keycloak/blob/a2c1055/services/src/main/java/org/keycloak/services/resources/admin/fgap/ClientPermissionsV2.java#L46-L48) returns false, and [`avoidClientIdPhishing`](https://github.com/keycloak/keycloak/blob/a2c1055/rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java#L72-L78) blocks before [`requireManage()`](https://github.com/keycloak/keycloak/blob/a2c1055/rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java#L186) ever runs.

tl;dr: `avoidClientIdPhishing` at line 157 was safe in V1 (role-based manage always implied list) but breaks in V2 (FGAP manage doesn't imply list).

Comment (vmuzikar) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: As discussed offline, for the scope of this PR, let's just skip the phishing check on POST. There's a follow-up for a broader fix: https://github.com/keycloak/keycloak/issues/47804

@mabartos Can you please update the PR?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 11 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:226]: @Pepo48 Expected. Related to this:
- https://github.com/keycloak/keycloak/issues/46553

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 6 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:247]: Hmm, not sure what's better. I'm still inclining to reuse the model for this to prevent any maintenance burden - as providing a new field to the representation, you need to keep track of places that touch that. 

@vmuzikar @Pepo48 WDYT? Or would you like to complete it in a follow-up?

File: rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java:70]: I guess it's related to your comment here: https://github.com/keycloak/keycloak/pull/47029#discussion_r3026457953

@Pepo48 @vmuzikar Are you ok with cleaning this up in the follow-up? Could you please create an issue?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 4 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:236]: @Pepo48 Commented on the other comment.

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 10 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: Updated.

**Comment by mabartos** [comment]
> During the review I noticed one more thing.

@Pepo48 Thanks for the detailed analysis! It'd be good if you could create follow-ups for your findings. WDYT?

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 6 above)
Comment (vmuzikar) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java:247]: Creating a model is far beyond ideal here. However, so is maintaining another mapping logic (v2 --> v1 rep). Let's keep as is and reassess in a follow-up.

File: rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java
Comment (vmuzikar) [reply rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/BaseClientModelMapper.java:70]: Let's tackle this under https://github.com/keycloak/keycloak/issues/47576.

**Comment by mabartos** [comment]
Rebased

File: rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java (hunk 9 above)
Comment (mabartos) [reply rest/admin-v2/api/src/main/java/org/keycloak/services/client/NewClientService.java]: Fixed

**Comment by Pepo48** [comment]
> It'd be good if you could create follow-ups for your findings. WDYT?

https://github.com/keycloak/keycloak/issues/47966

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks.

@Pepo48 Can you please re-review too?

**Review by Pepo48 — APPROVED** [review]
Thanks Martin!
